tests/frontend/test_circuitspecs_chip0.py

```diff
@@ -0,0 +1,300 @@
+# Copyright 2019 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Unit tests for the CircuitSpec class"""
+import textwrap
+
+import pytest
+import numpy as np
+
+import blackbird
+
+import strawberryfields as sf
+from strawberryfields.program_utils import CircuitError
+import strawberryfields.ops as ops
+from strawberryfields.io import to_program
+from strawberryfields.utils import random_interferometer
+from strawberryfields.circuitspecs.chip0 import Chip0Specs, CircuitSpecs
+
+
+pytestmark = pytest.mark.frontend
+
+np.random.seed(42)
+
+
+class DummyCircuit(CircuitSpecs):
+    """Dummy circuit used to instantiate
+    the abstract base class"""
+    modes = 4
+    remote = False
+    local = True
+    interactive = True
+    primitives = {"S2gate", "MeasureFock", "Rgate", "BSgate"}
+    decompositions = {"Interferometer": {}, "MZgate": {}}
+
+
+class TestChip0Compilation:
+    """Tests for compilation using the Chip0 circuit specification"""
+
+    def test_exact_template(self):
+        """Test compilation works for the exact circuit"""
+        bb = blackbird.loads(Chip0Specs.circuit)
+        bb = bb(
+            squeezing_amplitude_0=0.43,
+            squeezing_amplitude_1=0.65,
+            external_phase_0=0.54,
+            internal_phase_0=-0.23,
+            phi0=1.24,
+            phi1=-0.54,
+            phi2=4.12,
+            phi3=0,
+        )
+
+        expected = to_program(bb)
+        res = expected.compile("chip0")
+
+        for cmd1, cmd2 in zip(res.circuit, expected.circuit):
+            assert cmd1.op.__class__ == cmd2.op.__class__
+            assert [p.x for p in cmd1.op.p] == [p.x for p in cmd2.op.p]
+            assert [i.ind for i in cmd1.reg] == [i.ind for i in cmd2.reg]

```

Comment:
This is a really cumbersome way of checking the equality of two programs. A useful feature would be to overload the `Program.__eq__(self, other)` special method, in order to perform the `nx.is_isomorphic` check (similarly to that done in `CircuitSpecs.compile`, but also taking into account parameter values).

That would allow us to simply do `assert prog1 == prog2` in the tests.